### ISSUE TITLE
Rise windows together

### ISSUE THREAD

**Issue by rexackermann** [domferr/tilingshell#110]
It's kinda disturbing when i stack windows together and then I swich to another set of windows and then I have to do everything again
It would be nice if the window stack would act as one window while swiching windows.
Thank you, the ui is marvelous.

**Comment by domferr** [comment]
Hey, thank you for your suggestion! It seems a good idea but I'm not sure about the UI of this. Do you have any interesting idea or any example of other operating system or any GNOME extension already doing it?

**Comment by domferr** [comment]
We have got another issue (#33) about the same thing. I'm adding it here to document this

**Comment by rodrigomantoan** [comment]
I'd love to see that feature as well, I think [Tiling Assistant](https://github.com/Leleat/Tiling-Assistant) ([gnome page](https://extensions.gnome.org/extension/3733/tiling-assistant/)) does this nicely. 

I really like both extensions, and would love to see kind of an all-in-one haha. I'm using Tiling Shell, and I miss this from Tiling Assistant — and the option to pick apps for the other parts of the tiling _(when dragging/dropping to a tile position)_.

**Comment by Leleat** [comment]
Raising tiled windows as a group in Tiling Assistant is an all-or-nothing setting; i. e. you either enable the setting and _always_ raise all tiled windows in a group, or disable group-raising for all grouped tiled windows.

Recently I got a new laptop with Windows 11, so I looked at how they do it. Windows seems to do an optional "soft grouping". When using `Super` + `Tab` tiled windows will be shown as a grouped item but additionally they are also shown as separate windows. That way the user can choose to either raise the entire group or only individual windows. That might be a better handle it than the all-or-nothing setting of Tiling Assistant.

You could integrate this optional group-raising into the app switcher and the overview in GNOME. The only problem I see is: how can the group-raising be made optional when the user clicks a window on the desktop. Maybe the group-raising could be used when a modifier is held?

**Comment by rexackermann** [comment]
I actually got accustomed to this feature in tilling assistant as @rodrigomantoan mentioned. But that broke suddenly in my setup. Gnome breaks a lot of extensios after updates. My desktop still crashes every once in a while and I have no idea what causes it. > Hey, thank you for your suggestion! It seems a good idea but I'm not sure about the UI of this. Do you have any interesting idea or any example of other operating system or any GNOME extension already doing it?

**Comment by rexackermann** [comment]
There could be an icon to chose it maybe, or a left click right click thing ?

> You could integrate this optional group-raising into the app switcher and the overview in GNOME. The only problem I see is: how can the group-raising be made optional when the user clicks a window on the desktop. Maybe the group-raising could be used when a modifier is held?

**Comment by sekurilabs** [comment]
The way tilling assistant does it is that each layout is a group of windows. Then when you click on one application icon it raises all from that layout. This would mean that it would be needed to somehow catch which window is raised and to which layout (group) it belongs so they are then all raised at the same time.

**Comment by Leleat** [comment]
@rexackermann 

>  I actually got accustomed to this feature in tilling assistant as @rodrigomantoan mentioned. But that broke suddenly in my setup. Gnome breaks a lot of extensios after updates. My desktop still crashes every once in a while and I have no idea what causes it.

Well, that sounds like a bug and should be reported at Tiling Assistant's repo (if you are still using T-A) especially if you think the crash happens because of T-A. After all, bugs can't be fixed if they aren't known... and I, the maintainer of Tiling Assistant, haven't had any reports for this yet  😅

**Comment by rexackermann** [comment]
Sorry for not reporting, my setup was is a mess(so yes I was uncertain that it was tilling assistant's fault) and I had assumed it will be fixed after a couple of days after the gnome changes has been adopted.
Anyways, I just toogled it back on, no crashes this time. I will report to the repo if I spot any issue next time.

> @rexackermann 
> 
> >  I actually got accustomed to this feature in tilling assistant as @rodrigomantoan mentioned. But that broke suddenly in my setup. Gnome breaks a lot of extensios after updates. My desktop still crashes every once in a while and I have no idea what causes it.
> 
> Well, that sounds like a bug and should be reported at Tiling Assistant's repo (if you are still using T-A) especially if you think the crash happens because of T-A. After all, bugs can't be fixed if they aren't known... and I, the maintainer of Tiling Assistant, haven't had any reports for this yet  😅

**Comment by rodrigomantoan** [comment]
> Raising tiled windows as a group in Tiling Assistant is an all-or-nothing setting; i. e. you either enable the setting and always raise all tiled windows in a group, or disable group-raising for all grouped tiled windows.

Yeah, that's true, and it can be annoying sometimes, I think the modifier that you mentioned is great @Leleat. And the way Windows handles it makes a smooth experience in general.

> You could integrate this optional group-raising into the app switcher and the overview in GNOME. The only problem I see is: how can the group-raising be made optional when the user clicks a window on the desktop. Maybe the group-raising could be used when a modifier is held?

Could it be something toggled on settings? Example: enable group-raising by default or with a modifier because some would prefer group-raising by default, while others would prefer it not being the default.

**Comment by domferr** [comment]
> You could integrate this optional group-raising into the app switcher and the overview in GNOME. The only problem I see is: how can the group-raising be made optional when the user clicks a window on the desktop. Maybe the group-raising could be used when a modifier is held?

Hey @Leleat it is nice to see you here :heart: !
My opinion is all about UX: I believe that if a user clicks on a window it must raise that window only, however the user should have a way to click on a group of windows to raise all of them. Integrating this into the app switcher is good, as well as integrating into the overview!

- **Windows 11**: https://youtu.be/2DhxYHAjKXc?feature=shared&t=48 (at 0:48)
<p align="center"><img src="https://github.com/user-attachments/assets/d8f9efa5-a1f4-4b32-a621-18d92cc28a2b" width=400 /></p>

- **Chrome OS**: From my research they basically copied Windows 11 one. However, it seems to me they group windows in the overview instead 

<p align="center"><img src="https://github.com/user-attachments/assets/90d9042e-4f7b-4e16-8d60-05089b433033" width=400 /></p>

- **Mac OS**: I didn't find anything about this...

**Comment by Leleat** [comment]
> Hey @Leleat it is nice to see you here ❤️ !
My opinion is all about UX: I believe that if a user clicks on a window it must raise that window only, however the user should have a way to click on a group of windows to raise all of them. Integrating this into the app switcher is good, as well as integrating into the overview!

Thanks :heart:. A bit of backstory as to why I am here: T-A (Tiling Assistant) is sorta in maintenance mode, I only do some bug fixes and ports to new GNOME Shell versions. So I've been keeping an eye on Tiling Shell to see where it goes :upside_down_face:. If Tiling Shell implements some of the requested features that make the core experience of T-A (like this one or the window suggestions in #62), I might sunset T-A and Tiling Shell could offer a great migration path. Especially, since some features in Tiling Shell are already better than in T-A, for instance, T-A's layouts suck... Anyway, back on topic.

I agree with your point about the window raising. Among other things I regret the way I implemented it in T-A and that I didn't put more thought into it back then due to my lack of experience. One of the first feature request in T-A was to disable the "always-grouping" of tiled windows. I think there  were even users who reported the grouped-raising as a bug even though it was an actual feature :joy:. 

> Windows 11: https://youtu.be/2DhxYHAjKXc?feature=shared&t=48 (at 0:48)

In W11 the optional grouping is presented in the bottom panel - as shown in your linked video, in the overview, and in their window switcher. The grouping in the panel might be problematic for GNOME though since GNOME doesn't offer any previews in their Dash and popular extensions like Dash-to-Dock or Dash-to-Panel might conflict with custom implementation in Tiling Shell. Maybe some collaboration between extensions is needed here.

I don't know if you have access to W11 but in case you don't, I am going to provide some screenshots on how W11 integrates the grouping into their overview and window switcher.

## Windows 11

### Overview

<details>

<summary>Screenshot of Windows 11's overview</summary>

![Screenshot 2024-10-26 122230](https://github.com/user-attachments/assets/8d94dfd8-4cb8-49f1-a85c-55590a867078)
</details>

W11 simply adds an extra item for the tile group to the overview in addition to the items for the individual windows. One problem I see here is that the overview may get pretty cluttered and tile groups will be hard to find. Maybe tile groups could be put into a special row/spot or have some special styling... or maybe it's not that big of a deal.

### Window Switcher

<details>

<summary>Screen of Windows 11's window switcher</summary>

![Screenshot 2024-10-26 122559](https://github.com/user-attachments/assets/a96f22f6-8988-476c-82ec-4df850d1a00f)

</details>

In the window switcher it works pretty much the same. Windows adds an extra item for the tile group (the second item in my screenshot).

## macOS

I also have a macBook, so I can provide some reference for it, too :)

macOS had a pretty barebones (bad) tiling system. Until the release of macOS Sequoia last month, all tiled and "maximized" (technically fullscreened) windows were automatically put into separate workspaces. You couldn't have floating windows as part of that workspace. Those workspaces only allowed 1 maximized window or you could tile 2 windows side by side... and you had to have 2 windows tiled side by side. You couldn't just tile 1 window to 1 screen half. If you didn't choose another window to fill the rest of the space, the tiled window would automatically maximize (fullscreen) and fill the entire workspace. macOS calls this "split view". Here you can see it in action: https://www.youtube.com/watch?v=pBFkklUhH_U

This has gotten a bit better with macOS Sequoia. Split view still exists but they also added proper tiling... Although it's basically just a Windows 10 (not 11) clone. But it has no grouping. Windows resize and raise independently. They also aren't grouped in macOS's overview or app switcher. Here is a video that showcases some  of it: https://www.youtube.com/watch?v=GXych0RNQxk.

**Comment by domferr** [comment]
> Thanks ❤️. A bit of backstory as to why I am here: T-A (Tiling Assistant) is sorta in maintenance mode, I only do some bug fixes and ports to new GNOME Shell versions. So I've been keeping an eye on Tiling Shell to see where it goes 🙃. If Tiling Shell implements some of the requested features that make the core experience of T-A (like this one or the window suggestions in #62), I might sunset T-A and Tiling Shell could offer a great migration path. Especially, since some features in Tiling Shell are already better than in T-A, for instance, T-A's layouts suck... Anyway, back on topic.

Hey @Leleat I'm happy you are looking into Tiling Shell! I started this project for fun and T-A has always been my point of reference, it is an amazing work! The features you are talking about are in my roadmap, so I'm open to provide all the support needed :heart:. If you have already some ideas, feel free to reach me at anytime!

> In W11 the optional grouping is presented in the bottom panel - as shown in your linked video, in the overview, and in their window switcher. The grouping in the panel might be problematic for GNOME though since GNOME doesn't offer any previews in their Dash and popular extensions like Dash-to-Dock or Dash-to-Panel might conflict with custom implementation in Tiling Shell. Maybe some collaboration between extensions is needed here.

I'm replying late because I made some research in the last few days about this. I also came to the conclusion that some collaboration between extensions would be needed here!

Thank you for sharing all of this! Me, as a user, when I need to raise multiple windows, I do not use the overview. I use the window switcher instead, since it gets the same job done faster. I'm not sure it is just me or not :laughing:. However, having a group of windows in the overview would be so handy! It may be nice to have both solutions, imho.

However, as to my knowledge, creating groups of windows in the overview cannot be done without hacky workarounds...

**Comment by domferr** [comment]
Hi everyone, this is a quick update for you! I've been implementing this feature lately and I'm happy to share a first version of it 🥳

https://github.com/user-attachments/assets/56a7be50-d0a2-49cf-9b13-ab0a736a7213

Let me know what do you think about it. I'd appreciate your feedback, give it a try! Thank you 💟

- GNOME 42, 43 or 44: [GNOME.42-44.tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18812743/GNOME.42-44.tilingshell%40ferrarodomenico.com.zip)
- GNOME 45, 46, 47 or 48: [tilingshell@ferrarodomenico.com.zip](https://github.com/user-attachments/files/18812742/tilingshell%40ferrarodomenico.com.zip)

**Comment by domferr** [comment]
Hey @GuillaumeAmat thanks for asking! This is near to be finished and I'm planning to release it with the next update of Tiling Shell! Idk when, but probably in more than two weeks

**Comment by 3v1n0** [comment]
@domferr ok the group item is nice, however I still think that we should make possible these things (at least optionally):
 - Groups have priority over sigle windows in alt+tab (ordered using the activation time of the latest active windows of a group) 
 - selecting an app (via super+tab or dock) should still raise the group it belongs at first try
 - raising a window of a group by clicking on it (when another floating window is active) should always raise the group it belongs 

So... While raising a window of a group alone should be possible (via alt+tab or dock), it should only be a explicit action and a secondary choice, or we end up breaking the user tiling setup each time the focus changes

**Comment by domferr** [comment]
Now available! Enjoy! https://github.com/domferr/tilingshell/releases/tag/17.0 :star:
<div align="center">
  <a href="https://extensions.gnome.org/extension/7065/tiling-shell/" >
      <img src="https://img.shields.io/badge/Install%20from-extensions.gnome.org-4A86CF?style=for-the-badge&logo=Gnome&logoColor=white"/>
  </a>
</div>

**Comment by 3v1n0** [comment]
Cool, however it would be nice to optionally tie that behavior to the raising or a grouped window.

And in general to use the group last-active time, based on windows that compose it.

**Comment by domferr** [comment]
I'm not sure what the idea of a group is, but Tiling Shell doesn't have them. Basically everything that was tiled, will be raised. So yeah, the last tiled windows will be raised, unless you un-tiled them before

**Comment by domferr** [comment]
Hi everyone! Latest Tiling Shell version (v17.3) added an option to raise all the tiled windows together when raising one tiled window. Release notes are [here](https://github.com/domferr/tilingshell/releases/tag/17.3). Thanks

**Comment by 3v1n0** [comment]
Nice!

We should also discuss about an API to expose the tiles data for being used by other extensions.

In particular my idea is to show the tiling groups (as you do in alt-tab) in dash to dock, as a separate icon, similar to this...

![IMG_20260131_150130.jpg](https://github.com/user-attachments/assets/8e99fe81-bcfd-496c-899f-03bb7d436ce8)

**Comment by domferr** [comment]
Yes @3v1n0 this would be amazing! I abandoned this idea long time ago since I have no clue on how to add the group to the dock *from* Tiling Shell or how to influence Dash-to-dock to do so. Since you speak about an API, then I understand this is not possible and it must be done *from* Dash-to-dock.

Some ideas to brainstorm with:
- We could consider to expose the API via dbus?
- Tiling Shell could emit signals to which Dash-to-dock connects to?
- Tiling Shell is currently attaching tiling data to windows (if it is tiled and the tile coords and size), Dash-to-dock might already read those?

I created issue #496 for this

**Comment by 3v1n0** [comment]
No need to go through dbus, since we're both in the same process, we can just have a shared object and rely on it.

Something similar was done by desktop icons https://github.com/micheleg/dash-to-dock/blob/master/desktopIconsIntegration.js
